overlord/snapstate/policy/os.go

```diff
@@ -56,6 +61,19 @@ func (p *osPolicy) CanRemove(st *state.State, snapst *snapstate.SnapState, rev s
 		return errIsModel
 	}
 
+	var snapdState snapstate.SnapState
+	err := snapstate.Get(st, "snapd", &snapdState)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	// if snapd snap is not installed, then this might be a system that has
+	// received snapd updates via the core snap. in that case, we can't remove
+	// the core snap.
+	if !snapdState.IsInstalled() {
+		return errSnapdNotInstalled
```

Comment:
I added this new error type for this case, although I'm not sure if it is needed.

Reply:
As mentioned above I would leave a TODO here to indicate that on systems that do not re-execute it should be, in principle, fine to remove snapd snap as well.

Reply:
this check is relevant only if are removing the last revision of core present, that's why you had to change the other test